src/Hamiltonians/HubbardRealSpace.jl

```diff
@@ -228,9 +289,24 @@ function Base.show(io::IO, h::HubbardRealSpace)
     println(io, ")")
 end
 
+"""
+    ==(H::HubbardRealSpace, G::HubbardRealSpace)
+
+Overload equality due to stored potential energy arrays.
+"""
+Base.:(==)(H::HubbardRealSpace, G::HubbardRealSpace) = all(map(p -> getproperty(H, p) == getproperty(G, p), propertynames(H)))
+
 starting_address(h::HubbardRealSpace) = h.address
-diagonal_element(h::HubbardRealSpace, address) = local_interaction(address, h.u)
-diagonal_element(h::HubbardRealSpace{1}, address) = local_interaction(address, h.u[1])
+function diagonal_element(h::HubbardRealSpace, address)
+    int = iszero(h.u) ? 0. : local_interaction(address, h.u)
+    pot = iszero(h.v) ? 0. : trap_potential(address, h.potential)
+    return int + pot
+end
```

Comment:
Hmm, this looks like it would lead to branching in runtime, as `h.u` and `h.v` contain runtime information (about whether interactions are zero or not). This is not great as evaluating diagonal elements will occur very often.

I would make sense to inspect the code via `@code_llvm`, or via benchmarking.

One solution would be to elevate the `SArray`s `u` and `v` (and `t`) to type parameters, as was done in `HubbardMom1D`, for example. This should be doable, because `StaticArray`s are `isbits` data.

If the value of the interaction parameters is already known at compile time, then the compiler could optimise away the `if` condition in question. It would also put more strain on the compiler - and sometimes it just gives up. It is thus necessary to closely examine and measure whether it leads to efficient code.

Reply:
Another, maybe simpler solution (and easier on the compiler): If the harmonic trap, or the two-particle interactions are not used, the corresponding type parameters `V` and `U` could just be set to `Nothing`. Then in line 301 and 302 you could test on `U` and `V` being `Nothing`, which is compile-time resolvable.

This would require loosening the restrictions in lines  201 and 202, but these are not really needed.

Reply:
I prefer the `nothing` suggestion.

Reply:
I did benchmark it and I dont think there is a performance gain either way for the single component case. But it does improve it for multicomponent addresses. 

And it is a better way to do it.